### PR TITLE
Ignore JDBC Drivers artifacts

### LINKED ISSUE #22442: Ignore JDBC Drivers artifacts
### Description

We could ignore artifacts for JDBC drivers as we don't really need to have them during the runtime. Regarding the DB vendor option, we can keep only the relevant artifacts of a particular JDBC driver.

### PR DESCRIPTION
**Pull request by mabartos**
Closes #22442

### CODE REVIEW

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/IgnoredArtifacts.java
Comment (ahus1) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/IgnoredArtifacts.java:90]: Another optimization might be to remove also the JDBC drivers for each database as well, not only the Quarkus artifacts.

**Review by ghost — COMMENTED** [review]
Unreported flaky test detected, please review

**Comment by mabartos** [comment]
@ahus1 @vmuzikar Could you please check it? Thanks

**Review by ahus1 — COMMENTED** [review]
Thank you for this change. Will this prevent some of the warning messages we've seen earlier when a DB driver is removed? What are the performance benefits of this change - there is always the cost to make it more difficult to maintain. 

I could see the need for some users to change from an included DB driver to a self-provided DB driver. Would this change enable this, or is it unrelated? 

While the list of drivers looks correct to me, I leave the final call if the added complexity is smaller than the benefits that we would expect here. 

For anyone visiting this in the future, I'd like to see a at least a very short summary of the expected benefits.

**Comment by mabartos** [comment]
@ahus1 Thanks for your points. There are performance benefits, indeed. I'll describe it more tomorrow.

**Comment by mabartos** [comment]
This change brings quite a lot of benefits to us. First of all, the execution of build, static-init, and runtime build steps is done only for the specified DB. By this change, Quarkus is not aware of any other JDBC extension except the one for the chosen DB, so there's no need to execute these build steps, which could produce additional beans, service providers, configuration handling, agroal connection settings, and much more.

To summarize, we can prevent the execution of unused build steps, create additional resources, reduce memory allocations, GC executions, C1/C2 executions... 

I've made a brief analysis of the performance (few executions followed by a comparison):

**Startup**

- The significant improvement in startup is for AgroalProcessor as we saved ~80ms (cnt of CPU samples decreased by ~66%). 
- The overall improvement is ~100ms. 

**Heap** 
- I've encountered a small peek during the startup
- I assume objects provided during the build steps are destroyed after a few GC executions. 
- There's no peek anymore

Old


New


**GC** - GC executions reduced about ~20%.
**Compiler** - compilations reduced about ~13%

**IMPORTANT**: However, I expected significant changes around the class loading with the presence of the JDBC drivers' JARs. I think specifying the JDBC drivers into the ignored artifacts does not take any effect as it's probably used only for the custom Quarkus ClassLoaders anyway, which doesn't bring them into consideration if they do not have to be used. 

IMO, they can be removed from the codebase and reduce the need for manageability of it; I'll verify it.

@ahus1 
> Will this prevent some of the warning messages we've seen earlier when a DB driver is removed?

Yes, and much more as the build steps for different JDBC extensions are not executed at all.

> I could see the need for some users to change from an included DB driver to a self-provided DB driver. Would this change enable this, or is it unrelated?

I think it's unrelated to this PR as the logic behind using the custom JDBC drivers has not changed.

**Comment by mabartos** [comment]
@vmuzikar @ahus1 I've put some details in the comment above.

This PR is in a final state, and the JDBC Drivers' JARs have to be specified there.
IMO, it is definitely worth putting these changes into Keycloak despite the manageability it requires. 

@vmuzikar @ahus1 WDYT?

**Comment by ahus1** [comment]
@mabartos - looking at the images with with "new" using more memory than "old", it seems they have been swapped? 

With 100 ms improvement in built time, when we ask performance interested users to use pre-built images where possible, and ~7 MB of RAM during build time (which is not relevant at run time), I'm still not convinced. 

At the same time, I won't veto it if the CND still wants to merge it.

**Comment by ahus1** [comment]
@mabartos - thank you for the additional explanations. No need for meeting, let's wait for @vmuzikar's response only.

File: quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/IgnoredArtifactsTest.java
Comment (vmuzikar) [inline quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/IgnoredArtifactsTest.java:118]: I think we could also check the unused DB artifacts are not present.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@mabartos Thanks for the PR!

The ignored artifacts should really affect only build time. They are in the end mapped to [`quarkus.class-loading.removed-artifacts`](https://quarkus.io/guides/class-loading-reference#quarkus-class-loading-configuration-class-loading-config_quarkus.class-loading.removed-artifacts) which is build time option. If I understand correctly, you were testing runtime with auto-build enabled. Is that correct? I'd like to see if there's any performance improvement in case of optimized image, i.e. if we omit the build phase.

What slightly concerns me is that it appears after this change more memory is allocated (at least initially until things settle down)? I know it removed the initial spike but the overall memory is higher?

@pedroigor Do you want to review as well?

**Comment by mabartos** [comment]
Okay, it seems I haven't explained it accurately, as it may cause some confusion around it; my bad. 

I'll answer the questions here and probably summarize the provided changes and benefits in a more clear way. Probably even start the Keycloak Benchmark, if it provides the functionality to compare distros with/without these changes (like startup time, memory footprint at the start, GC activities,...?)

> The ignored artifacts should really affect only build time. They are in the end mapped to [quarkus.class-loading.removed-artifacts](https://quarkus.io/guides/class-loading-reference#quarkus-class-loading-configuration-class-loading-config_quarkus.class-loading.removed-artifacts) which is build time option.

@vmuzikar
Even when it's a build time option, it doesn't mean it doesn't affect runtime, of course. Quite the opposite; that's why we can execute the augmented application. The ignored artifacts are banned for all Quarkus ClassLoaders, so the Base Runtime and Runtime ClassLoader as well. It means they're completely removed from the application execution. To be more specific, no beans are created in static-init/runtime-init bootstrap phase defined by these extensions, ServiceLoaders do not manage any provider included in these JARs,...

> If I understand correctly, you were testing runtime with auto-build enabled. Is that correct? I'd like to see if there's any performance improvement in case of optimized image, i.e. if we omit the build phase.

Not at all. There wasn't any build phase done during the testing; the optimized version was used.

> What slightly concerns me is that it appears after this change more memory is allocated (at least initially until things settle down)? I know it removed the initial spike but the overall memory is higher?

Nope, the overall memory was quite the same after some time. These pictures were only for demonstrating the initial peak. Described in comment https://github.com/keycloak/keycloak/pull/22443#issuecomment-1723624265.

**Comment by mabartos** [comment]
This PR is starting to be quite messy as there are certain concerns about perf improvements it might bring vs. manageability.

I'll try to find out better approach to gathering performance results to emphasize it more :stuck_out_tongue_closed_eyes: (will try to leverage Keycloak Benchmark)

File: quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/IgnoredArtifactsTest.java
Comment (mabartos) [reply quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/IgnoredArtifactsTest.java:118]: Ok

**Comment by abstractj** [comment]
@mabartos based on the discussions here I believe we have a couple of action items to follow up from my perspective (maybe others will disagree):

1. Address flaky tests: investigate the unreported flaky tests mentioned in the pull request to determine if they are affected by the changes proposed.
2. Clarity about the performance benefits: provide a more detailed analysis of the performance improvements, highlighting the specific improvements in startup time, memory allocations, GC executions...
3. Run tests with auto-build enabled to determine if there are any performance improvements in the case of an optimized image.

@vmuzikar do the action items above make any sense?

**Comment by mabartos** [comment]
# Performance analysis
This performance analysis touches on multiple aspects of the application's performance. JFR (Java Flight Recorder), Async profiler, and Podman stats, for testing performance for optimized Keycloak image, were the main tools used for the performance analysis. 

The expected improvement is related to the ignoring JARs for JDBC drivers and its related Quarkus extensions. We will prevent additional Class loading, C1/C2 executions, GC cycles, and memory allocations.

TERMINOLOGY: In this analysis, I'll compare behavior before and after these changes. They will be marked as **OLD** and **NEW**.

## Profiling
In this section, I'll describe the highlights of benefits these changes bring to us analyzed with the usage of JFR and Async profiler. 

### CPU
I had the possibility to test it on two different machines as we can also see the difference between more and less performant agents. In the tables below, there are described main differences in used CPU samples during the profiling. CPU samples basically represent a count of encountered snapshots in time, where we can find out what operations' execution occupies the CPU the most.

CPU sampling 1ms/sample. The numbers in the table below represent a count of CPU samples. 

**Lenovo T490s (Intel Core i7 Pro 8xCPU, 16GB Mem)**
<table>
    <tr>
       <td align="center">Method</td>
       <td align="center">OLD</td>
       <td align="center">NEW</td>
       <td align="center">DIFF</td>
    </tr>
    <tr>
      <td>AgroalProcessor.generateDataSourceBeans</td>
      <td>163</td>
      <td>34</td>
      <td>-129</td>
    </tr>
    <tr>
      <td>  --- (inner) DataSources.loadDriversInTCCL</td>
      <td>125</td>
      <td>2</td>
      <td>-123</td>
    </tr>
     <tr>
      <td>HibernateOrmRecorder.created</td>
      <td>748</td>
      <td>712</td>
      <td>-36</td>
    </tr>
    <tr>
      <td>DataSource.init</td>
      <td>15</td>
      <td>11</td>
      <td>-4</td>
    </tr>
  </table>
  
**Lenovo X1 Nano (12th gen Intel Core i7-1280P 20xCPU, 32GB Mem)**
<table>
    <tr>
       <td align="center">Method</td>
       <td align="center">OLD</td>
       <td align="center">NEW</td>
       <td align="center">DIFF</td>
    </tr>
    <tr>
      <td>AgroalProcessor.generateDataSourceBeans</td>
      <td>70</td>
      <td>20</td>
      <td>-50</td>
    </tr>
    <tr>
      <td>  --- (inner) DataSources.loadDriversInTCCL</td>
      <td>53</td>
      <td>1</td>
      <td>-52</td>
    </tr>
     <tr>
      <td>HibernateOrmRecorder.created</td>
      <td>388</td>
      <td>345</td>
      <td>-43</td>
    </tr>
    <tr>
      <td>DataSource.init</td>
      <td>15</td>
      <td>11</td>
      <td>-4</td>
    </tr>
  </table>  
  
### Memory allocations
The biggest difference in memory allocations is in the `AgroalProcessor` as well. As inside of the method, the call for `DataSources.loadDriversInTCCL()` is done and it loads all JDBC drivers found by ServiceLoader. As we will load only the specific JDBC driver we want, the memory allocation for it is drastically decreased.

<table>
    <tr>
       <td align="center">Method</td>
       <td align="center">OLD</td>
       <td align="center">NEW</td>
       <td align="center">DIFF</td>
    </tr>
    <tr>
      <td>AgroalProcessor.generateDataSourceBeans</td>
      <td>1.7 MB</td>
      <td>9.5 MB</td>
      <td>-7.8 MB</td>
    </tr>
  </table>
  
### Startup time
The biggest improvement in the startup time is also related to the AgroalProcessor, as the CPU samples decreased.  
 
 <table>
    <tr>
       <td align="center">Build item</td>
       <td align="center">OLD</td>
       <td align="center">NEW</td>
       <td align="center">DIFF</td>
    </tr>
    <tr>
      <td>AgroalProcessor.generateDataSourceBeans</td>
      <td>124 ms</td>
      <td>23 ms</td>
      <td>-101 ms</td>
    </tr>
    <tr>
      <td>ArcProcessor.generateResources</td>
      <td>456 ms</td>
      <td>414 ms</td>
      <td>-42 ms</td>
    </tr>
  </table>
  
  NOTE: Only approximately; it differs execution by execution, but most of the time very similar. 
  
  ### Other improvements
I've spotted also other improvements in different areas of the application and I can share some of them here. I haven't encountered any negative performance impact provided by these changes.

 <table>
    <tr>
       <td align="center">Aspects</td>
       <td align="center">OLD</td>
       <td align="center">NEW</td>
       <td align="center">DIFF</td>
    </tr>
    <tr>
      <td>CPU Load Max</td>
      <td>43.4 %</td>
      <td>40.2 %</td>
      <td>-3.2 %</td>
    </tr>
    <tr>
      <td>Loaded class count</td>
      <td>19049</td>
      <td>18763</td>
      <td>-286</td>
    </tr>
    <tr>
      <td>Quarkus Class Loader Chunk size</td>
      <td>10.4 MiB</td>
      <td>10.3 MiB</td>
      <td>-0.1 MiB</td>
    </tr>
    <tr>
      <td>Garbage Collection count</td>
      <td>26</td>
      <td>25</td>
      <td>-1</td>
    </tr>
    <tr>
      <td>Garbage Collection duration (ParallelGC)</td>
      <td>39.089 ms</td>
      <td>33.197 ms</td>
      <td>-5.89 ms</td>
    </tr>
  </table>
  
## Optimized image
In this section, I'll describe the highlights of benefits these changes bring to us analyzed by Podman statistics (`podman stats`) for the optimized Keycloak image. The build phase was executed in the _builder_ phase of a Dockerfile and tested was only the runtime behavior for the optimized image with H2 in-memory DB without executing any DB schema init or upgrade. 

#### Benchmark settings

- Analyzing CPU utilization and Memory usage
- Snapshots taken every 1s (`podman stats --no-reset -i 1`)
- Optimized Keycloak images
- **OLD** (Keycloak main) and **NEW** (this PR) image was executed 3x 
- Computed average of these values for every snapshot

#### Memory usage
You can noticed the memory usage is lower for the NEW image and after ~10s, it's more or less stabilized as the old one. 
We don't encounter the initial peak for the memory usage due to the JDBC Driver JARs inits. No comprehensive GC cycles.
The memory usage is related to the whole image; with JVM. 



#### Average CPU Usage
The average CPU is the same in ~0-2s, but then, I assume, the JDBC Drivers loading with followed GC execution is done and the average CPU usage is higher for the OLD image. For more info how the Average CPU is computed for the Podman/Docker stats may be found here[1]. 



[1] https://stackoverflow.com/questions/47401648/docker-stats-with-cpu-percentage-more-than-100

## Testing device info
```
CPU(s):                  20
  On-line CPU(s) list:   0-19
Vendor ID:               GenuineIntel
  Model name:            12th Gen Intel(R) Core(TM) i7-1280P
    CPU family:          6
    Model:               154
    Thread(s) per core:  2
    Core(s) per socket:  14
    Socket(s):           1
    Stepping:            3
    CPU(s) scaling MHz:  31%
    CPU max MHz:         4800.0000
    CPU min MHz:         400.0000
Memory:                  32GB    
```

## Summary 

@vmuzikar @ahus1 @abstractj
Here is the detailed performance analysis for the benefits it will brings to us. I hope it's sufficient to show that it'll improve the overall performance. Please let me know, if you're more curious about it, you want the testing data or you have some input on this :))

I'd be glad to answer your questions and react to your comments. 

I think it worths to include it into Keycloak even there will be slight need for maintanance. WDYT?

**Comment by mabartos** [comment]
@abstractj Actions items summary:

- [ ] Address flaky tests; Flaky tests are not related to this PR.
- [ ] Clarity about the performance benefits; Provided in the comment above.
- [ ] Run tests with auto-build enabled; part of the performance analysis above.

**Comment by ahus1** [comment]
I abstain from voting here, I leave it to the CND team to decide.

**Comment by mabartos** [comment]
I just wanted to add my thought to the certain performance/maintainability controversion. 

What's the worst that can happen? If some of the transitive dependencies (the JDBC driver JAR) change in the Quarkus extension and it's not reflected here, it'll just not be ignored and will be processed. So, we would end up in the same state as we're now in Keycloak main ;)

But there's a tiny probability it'll happen, as it'd be reflected in the Quarkus release notes, and it will not be changed so often anyway(if almost ever). 

It just seems to me weird that if a user wants to use Postgres DB, the stuff related to Oracle, MS-SQL, MariaDB,... is processed. We should avoid such situations.

I'm a little bit surprised it unleashed such a broad conversation about it and took a considerable amount of effort to promote it :) 

@vmuzikar @pedroigor I'd be glad for your input.

**Review by vmuzikar — APPROVED** [review]
@mabartos Thanks for the elaborate analysis.

While the perf difference not that impressive from my perspective, I see relatively low risk in introducing this optimization with acceptable maintenance costs of it.

Of course it's not ideal as the final goal here is to have some more generic capability to exclude unwanted/unused bits from runtime and even the distribution based on selected features. But that's certnainly not something we can prioritize now.

I'm going to approve but I'd like to get a green tick from @pedroigor as well was involved in the discussion too, IIRC.

**Review by pedroigor — APPROVED** [review]

**Review by shawkins — APPROVED** [review]

**Pull request merged**
